doc/mgr/prometheus.rst

```diff
@@ -79,62 +79,56 @@ drive statistics, special series are output like this:
-    ceph_disk_occupation{ceph_daemon="osd.0",device="sdd",instance="myhost",job="ceph"}
+    ceph_disk_occupation{ceph_daemon="osd.0",device="sdd", hostname="myhost"}
 
-To use this to get disk statistics by OSD ID, use the ``and on`` syntax
-in your prometheus query like this:
+To use this to get disk statistics by OSD ID, use either the ``and`` operator or
+the ``*`` operator in your prometheus query. All metadata metrics (like ``
+ceph_disk_occupation`` have the value 1 so they act neutral with ``*``. Using ``*``
+allows to use ``group_left`` and ``group_right`` grouping modifiers, so that
+the resulting metric has additional labels from one side of the query.
+
+See the
+`prometheus documentation`__ for more information about constructing queries.
+
+__ https://prometheus.io/docs/prometheus/latest/querying/basics
+
+The goal is to run a query like
 
 ::
 
     rate(node_disk_bytes_written[30s]) and on (device,instance) ceph_disk_occupation{ceph_daemon="osd.0"}
 
-See the prometheus documentation for more information about constructing
-queries.
-
-Note that for this mechanism to work, Ceph and node_exporter must agree
-about the values of the ``instance`` label.  See the following section
-for guidance about to to set up Prometheus in a way that sets
-``instance`` properly.
+Out of the box the above query will not return any metrics since the ``instance`` labels of
+both metrics don't match. The ``instance`` label of ``ceph_disk_occupation``
+will be the currently active MGR node. Conversely the query
 
-Configuring Prometheus server
-=============================
+::
 
-See the prometheus documentation for full details of how to add
-scrape endpoints: the notes
-in this section are tips on how to configure Prometheus to capture
-the Ceph statistics in the most usefully-labelled form.
+    rate(node_disk_bytes_written[30s]) and on (device,hostname) ceph_disk_occupation{ceph_daemon="osd.0"}
 
-This configuration is necessary because Ceph is reporting metrics
-from many hosts and services via a single endpoint, and some
-metrics that relate to no physical host (such as pool statistics).
+won't return any metrics since the node_exporter does not attach a ``hostname``
+label. The following two section outline two approaches to remedy this.
 
-honor_labels
-------------
+Use label_replace
+=================
 
-To enable Ceph to output properly-labelled data relating to any host,
-use the ``honor_labels`` setting when adding the ceph-mgr endpoints
-to your prometheus configuration.
+The ``label_replace`` function (cp.
+`label_replace documentation <https://prometheus.io/docs/prometheus/latest/querying/functions/#label_replace>`_)
+can add a label to, or alter a label of, a metric within a query.
 
-Without this setting, any ``instance`` labels that Ceph outputs, such
-as those in ``ceph_disk_occupation`` series, will be overridden
-by Prometheus.
+To correlate an OSD and its disks write rate, the following query can be used:
 
-Ceph instance label
```

Comment:
Why does the "Ceph instance label" section go away?  I don't see how the changes here would affect the way stats end up with an "instance" label by default when scraped, which will change (undesirable) depending on which mgr is active

Reply:
I removed this section along with the "honor_labels" section. I'd argue that one actually wants the correct instance labels. The info which mgr is active is otherwise lost.

in queries one can use the ```=~``` label matcher to include all mgr instances. For example:
```
ceph_osd_up{instance=~".*:9283"}
``` 
returns data no matter which mgr is active. In Grafana dashboards can use this matching style in conjunction with dashboard templates.

Reply:
Point taken regarding multi-cluster environments. How about I revise this section to document both the instance re-label approach and the additional label approach?

> Keeping the mgr hostname out of the labels isn't just about queryability, it's also about avoiding bloating the DB with N different time series for each counter for N mgr daemons.

Do we know this, I'd actually be curious what overhead this incurs. I would argue that this shouldn't be an issue since there is always only one mgr exporting metrics. I would hope that the overhead of say three timeseries of length 1 is negligible compared to one timeseries of length 3 (given otherwise more or less the same data/labels). As far as I understand this is exactly what prometheus does well.

Reply:
I haven't looked into it in detail (working from vague recollections of presentations about the data format), but I wouldn't assume that the impact of having 3x time series would be negligible -- there will only be data in any one of them at any given time offset, but some indices may be 3x bigger in any range where multiple mgrs were outputting.

As in my other comment at the end of the PR, I'm kind of unclear on what's preferable about using hostname= rather than instance= here?